File: server/routes/clients.ts
import { Router } from 'express';
import { isAuthenticated } from '../replitAuth';
import { storage } from '../storage';
import { insertClientSchema } from '@shared/schema';

const router = Router();

// Rota para buscar todos os clientes (com limite)
router.get('/', isAuthenticated, async (req, res) => {
  try {
    const limit = req.query.limit ? parseInt(req.query.limit as string) : undefined;
    const userId = req.user?.claims?.sub || "999999"; // Fallback para testes
    
    if (!userId) {
      return res.status(401).json({
        success: false,
        message: 'ID de usuário não encontrado'
      });
    }
    
    const clients = await storage.getUserClients(userId, limit);
    
    res.json({
      success: true,
      clients
    });
  } catch (error) {
    console.error('Erro ao buscar clientes:', error);
    res.status(500).json({
      success: false,
      message: 'Erro ao buscar clientes'
    });
  }
});

// Rota para buscar um cliente específico
router.get('/:id', isAuthenticated, async (req, res) => {
  try {
    const clientId = parseInt(req.params.id);
    
    if (isNaN(clientId)) {
      return res.status(400).json({
        success: false,
        message: 'ID de cliente inválido'
      });
    }
    
    const client = await storage.getClient(clientId);
    
    if (!client) {
      return res.status(404).json({
        success: false,
        message: 'Cliente não encontrado'
      });
    }
    
    res.json({
      success: true,
      client
    });
  } catch (error) {
    console.error('Erro ao buscar cliente:', error);
    res.status(500).json({
      success: false,
      message: 'Erro ao buscar cliente'
    });
  }
});

// Rota para criar um novo cliente
router.post('/', isAuthenticated, async (req, res) => {
  try {
    const userId = req.user?.claims?.sub || "999999"; // Fallback para testes
    
    if (!userId) {
      return res.status(401).json({
        success: false,
        message: 'ID de usuário não encontrado'
      });
    }
    
    const validationResult = insertClientSchema.safeParse(req.body);
    
    if (!validationResult.success) {
      return res.status(400).json({
        success: false,
        message: 'Dados de cliente inválidos',
        errors: validationResult.error.errors
      });
    }
    
    const clientData = {
      ...validationResult.data,
      userId
    };
    
    const client = await storage.createClient(clientData);
    
    res.status(201).json({
      success: true,
      client
    });
  } catch (error) {
    console.error('Erro ao criar cliente:', error);
    res.status(500).json({
      success: false,
      message: 'Erro ao criar cliente'
    });
  }
});

// Rota para atualizar um cliente
router.patch('/:id', isAuthenticated, async (req, res) => {
  try {
    const clientId = parseInt(req.params.id);
    
    if (isNaN(clientId)) {
      return res.status(400).json({
        success: false,
        message: 'ID de cliente inválido'
      });
    }
    
    const existingClient = await storage.getClient(clientId);
    
    if (!existingClient) {
      return res.status(404).json({
        success: false,
        message: 'Cliente não encontrado'
      });
    }
    
    const userId = req.user?.claims?.sub || "999999"; // Fallback para testes
    if (existingClient.userId !== userId) {
      return res.status(403).json({
        success: false,
        message: 'Não autorizado a modificar este cliente'
      });
    }
    
    // Aceita atualizações parciais
    const updatedClient = await storage.updateClient(clientId, req.body);
    
    res.json({
      success: true,
      client: updatedClient
    });
  } catch (error) {
    console.error('Erro ao atualizar cliente:', error);
    res.status(500).json({
      success: false,
      message: 'Erro ao atualizar cliente'
    });
  }
});

// Rota para excluir um cliente
router.delete('/:id', isAuthenticated, async (req, res) => {
  try {
    const clientId = parseInt(req.params.id);
    
    if (isNaN(clientId)) {
      return res.status(400).json({
        success: false,
        message: 'ID de cliente inválido'
      });
    }
    
    const existingClient = await storage.getClient(clientId);
    
    if (!existingClient) {
      return res.status(404).json({
        success: false,
        message: 'Cliente não encontrado'
      });
    }
    
    const userId = req.user?.claims?.sub || "999999"; // Fallback para testes
    if (existingClient.userId !== userId) {
      return res.status(403).json({
        success: false,
        message: 'Não autorizado a excluir este cliente'
      });
    }
    
    const success = await storage.deleteClient(clientId);
    
    if (!success) {
      return res.status(500).json({
        success: false,
        message: 'Falha ao excluir cliente'
      });
    }
    
    res.json({
      success: true,
      message: 'Cliente excluído com sucesso'
    });
  } catch (error) {
    console.error('Erro ao excluir cliente:', error);
    res.status(500).json({
      success: false,
      message: 'Erro ao excluir cliente'
    });
  }
});

export default router;

File: app/api/endpoints/clients.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from app.db.session import get_db
from app.models.user import User
from app.models.client import Client
from app.schemas.client import ClientCreate, ClientUpdate, Client as ClientSchema, ClientList
from app.utils.security import get_current_user
from app.utils.logger import logger

router = APIRouter()

@router.get("", response_model=ClientList)
async def get_clients(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
    skip: int = 0,
    limit: int = 50
):
    """
    Obtém a lista de clientes do usuário atual
    """
    clients = db.query(Client).filter(Client.user_id == current_user.id).offset(skip).limit(limit).all()
    return {"clients": clients}

@router.get("/{client_id}", response_model=ClientSchema)
async def get_client(
    client_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Obtém um cliente específico por ID
    """
    client = db.query(Client).filter(Client.id == client_id).first()
    
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Cliente não encontrado"
        )
    
    # Verificar se o cliente pertence ao usuário atual
    if client.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Sem permissão para acessar este cliente"
        )
    
    return client

@router.post("", response_model=ClientSchema)
async def create_client(
    client_create: ClientCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Cria um novo cliente
    """
    # Definir o user_id como o ID do usuário atual
    client_data = client_create.dict()
    client_data["user_id"] = current_user.id
    
    client = Client(**client_data)
    
    try:
        db.add(client)
        db.commit()
        db.refresh(client)
        return client
    except Exception as e:
        db.rollback()
        logger.error(f"Erro ao criar cliente: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Erro ao salvar o cliente"
        )

@router.put("/{client_id}", response_model=ClientSchema)
async def update_client(
    client_id: int,
    client_update: ClientUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Atualiza um cliente existente
    """
    client = db.query(Client).filter(Client.id == client_id).first()
    
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Cliente não encontrado"
        )
    
    # Verificar se o cliente pertence ao usuário atual
    if client.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Sem permissão para editar este cliente"
        )
    
    # Atualizar campos que foram fornecidos
    update_data = client_update.dict(exclude_unset=True)
    for key, value in update_data.items():
        setattr(client, key, value)
    
    try:
        db.commit()
        db.refresh(client)
        return client
    except Exception as e:
        db.rollback()
        logger.error(f"Erro ao atualizar cliente: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Erro ao atualizar o cliente"
        )

@router.delete("/{client_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_client(
    client_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Exclui um cliente
    """
    client = db.query(Client).filter(Client.id == client_id).first()
    
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Cliente não encontrado"
        )
    
    # Verificar se o cliente pertence ao usuário atual
    if client.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Sem permissão para excluir este cliente"
        )
    
    try:
        db.delete(client)
        db.commit()
    except Exception as e:
        db.rollback()
        logger.error(f"Erro ao excluir cliente: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Erro ao excluir o cliente"
        )